phpBB/phpbb/module/module_manager.php

```
@@ -0,0 +1,552 @@
+
+		// Resync tree
+		$sql = 'UPDATE ' . $this->modules_table . "
+			SET right_id = right_id - $diff
+			WHERE module_class = '" . $this->db->sql_escape($module_class) . "'
+				AND left_id < {$row['right_id']} AND right_id > {$row['right_id']}";
+		$this->db->sql_query($sql);
+
+		$sql = 'UPDATE ' . $this->modules_table . "
+			SET left_id = left_id - $diff, right_id = right_id - $diff
+			WHERE module_class = '" . $this->db->sql_escape($module_class) . "'
+				AND left_id > {$row['right_id']}";
+		$this->db->sql_query($sql);
+
+		return array();
+	}
+
+	/**
+	 * Move module position by $steps up/down
+	 *
+	 * @param array		$module_row		Array of module data
+	 * @param string	$module_class	Class of the module (acp, ucp, mcp etc...)
+	 * @param string	$action			Direction of moving (valid values: move_up or move_down)
+	 * @param int		$steps			Number of steps to move module
+	 *
+	 * @return string
+	 */
+	public function move_module_by($module_row, $module_class, $action = 'move_up', $steps = 1)
+	{
+		/**
+		 * Fetch all the siblings between the module's current spot
+		 * and where we want to move it to. If there are less than $steps
+		 * siblings between the current spot and the target then the
+		 * module will move as far as possible
+		 */
+		$sql = 'SELECT module_id, left_id, right_id, module_langname
+			FROM ' . $this->modules_table . "
+			WHERE module_class = '" . $this->db->sql_escape($module_class) . "'
+				AND parent_id = " . (int) $module_row['parent_id'] . '
+				AND ' . (($action == 'move_up') ? 'right_id < ' . (int) $module_row['right_id'] . ' ORDER BY right_id DESC' : 'left_id > ' . (int) $module_row['left_id'] . ' ORDER BY left_id ASC');
+		$result = $this->db->sql_query_limit($sql, $steps);
+
+		$target = array();
+		while ($row = $this->db->sql_fetchrow($result))
+		{
+			$target = $row;
+		}
+		$this->db->sql_freeresult($result);
+
+		if (!sizeof($target))
+		{
+			// The module is already on top or bottom
+			return false;
```

I think we should throw an exception, instead of mixed returns?